pkg/controllers/override/imageparse.go

```diff
@@ -0,0 +1,175 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package override
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/distribution/reference"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	utilreference "github.com/kubewharf/kubeadmiral/pkg/lifted/distribution/reference"
+)
+
+type Image struct {
+	registry   string
+	repository string
+	tag        string
+	digest     string
+}
+
+func (i *Image) OperateRegistry(operator, value string) {
+	switch operator {
+	case OperatorAdd:
+		i.registry += value
+	case OperatorReplace:
+		i.registry = value
+	case OperatorRemove:
+		i.registry = ""
+	}
+}
+
+func (i *Image) OperateRepository(operator, value string) {
+	switch operator {
+	case OperatorAdd:
+		i.repository += value
+	case OperatorReplace:
+		i.repository = value
+	case OperatorRemove:
+		i.repository = ""
+	}
+}
+
+func (i *Image) OperateTag(operator, value string) {
+	var newTag string
+	switch operator {
+	case OperatorAdd:
+		newTag = i.tag + value
+	case OperatorReplace:
+		newTag = value
+	case OperatorRemove:
+		newTag = ""
+	}
+
+	if operator != OperatorRemove && !utilreference.AnchoredTagRegexp.MatchString(newTag) {
+		return
+	}
+
+	i.tag = newTag
+}
+
+func (i *Image) OperateDigest(operator, value string) {
+	var newDigest string
+	switch operator {
+	case OperatorAdd:
+		newDigest = i.digest + value
+	case OperatorReplace:
+		newDigest = value
+	case OperatorRemove:
+		newDigest = ""
+	}
+
+	if operator != OperatorRemove && !utilreference.AnchoredDigestRegexp.MatchString(newDigest) {
+		return
+	}
+
+	i.digest = newDigest
+}
+
+func (i *Image) String() string {
+	fullName := i.repository
+
+	if i.registry != "" {
+		fullName = strings.Join([]string{i.registry, i.repository}, "/")
+	}
+
+	if i.tag != "" {
+		fullName = strings.Join([]string{fullName, i.tag}, ":")
+	}
+
+	if i.digest != "" {
+		fullName = strings.Join([]string{fullName, i.digest}, "@")
+	}
+
+	return fullName
+}
+
+// ParseImage returns a Components of the given image.
+func ParseImage(fullName string) (*Image, error) {
+	ref, err := reference.Parse(fullName)
+	if err != nil {
+		return nil, err
+	}
+
+	image := &Image{}
+	if named, ok := ref.(reference.Named); ok {
+		image.registry, image.repository = SplitRegistryAndRepository(named.Name())
+	}
+
+	if tagged, ok := ref.(reference.Tagged); ok {
+		image.tag = tagged.Tag()
+	}
+
+	if digested, ok := ref.(reference.Digested); ok {
+		image.digest = digested.Digest().String()
+	}
+
+	return image, nil
+}
+
+func SplitRegistryAndRepository(name string) (registry, repository string) {
+	i := strings.IndexRune(name, '/')
+	if i == -1 || (!strings.ContainsAny(name[:i], ".:") && name[:i] != "localhost") {
```

Comment:
Is there a reference for this logic?

Reply:
The grammar can be found [here](https://github.com/distribution/reference/blob/49c28499d219290c3226822e9cfcd4ede6d75379/reference.go#L1). 

I think the `[Domain](https://github.com/distribution/reference/blob/49c28499d219290c3226822e9cfcd4ede6d75379/reference.go#L151)` and `[Path](https://github.com/distribution/reference/blob/49c28499d219290c3226822e9cfcd4ede6d75379/reference.go#L160)` functions provide the functionality we need here (correct me if I'm wrong). 

I would prefer if we use their implementation and not reinvent the wheel.

Reply:
the reference is here: https://github.com/distribution/distribution/blob/v2.7.1/reference/normalize.go#L62

Reply:
The reason for using reference here is due to this discussion: https://stackoverflow.com/a/37867949

If we use `image.registry, image.repository = named.Domain(), named.Path()` for parsing, the following problems will occur: `ealen/echo-server` is resolved to `registry: "ealen", repository : "echo-server"`.

Using the function in the reference ensures that the above image is parsed into the following content: `registry: "", repository: "ealen/echo-server"`

Reply:
The comment has been added.